### PR TITLE
Update GLFW embedding to support AOT mode

### PR DESCRIPTION
**Pull request by MarcusTomlinson**
Unblocks: https://github.com/flutter/flutter/pull/57135

Together with the above version of flutter tool, one can test this by building and running any application in profile or release mode like so:

`flutter --local-engine=host_profile_unopt build linux --profile`
or
`flutter --local-engine=host_release_unopt build linux --release`

### CODE REVIEW

File: shell/platform/glfw/flutter_glfw.cc [hunk 1]
```diff
@@ -618,6 +628,22 @@ static bool RunFlutterEngine(
   args.command_line_argv = &argv[0];
   args.platform_message_callback = EngineOnFlutterPlatformMessage;
   args.custom_task_runners = &task_runners;
+
+  if (FlutterEngineRunsAOTCompiledDartCode()) {
```
Comment (stuartmorgan) [inline shell/platform/glfw/flutter_glfw.cc:632]: There's already a lot going on in this function, and this is non-trivial code, so I'd strongly prefer it be extracted to a method (see https://github.com/flutter/engine/pull/18589 where I stole this, but made it a method :) )

File: shell/platform/glfw/flutter_glfw.cc [hunk 2]
```diff
@@ -618,6 +628,22 @@ static bool RunFlutterEngine(
...
+    FlutterEngineAOTDataSource data_in;
+    FlutterEngineAOTData data_out;
```
Comment (stuartmorgan) [inline shell/platform/glfw/flutter_glfw.cc:634]: See Chinmay's comments about initialization in the review of my PR linked above.

In fact, it's probably easiest to just steal back the modified version and use that. You can remove the `u8string` stuff though, since that's Windows-specific.

File: shell/platform/glfw/flutter_glfw.cc [hunk 3]
```diff
@@ -618,6 +628,22 @@ static bool RunFlutterEngine(
...
+    std::string elf_path = assets_path_string + "/libapp.so";
```
Comment (stuartmorgan) [inline shell/platform/glfw/flutter_glfw.cc:635]: Per my comment in the other issue, I think you should add a new API parameter instead of having this be relative to assets_path.

File: shell/platform/glfw/flutter_glfw.cc (hunk 2 above)
Comment (MarcusTomlinson) [reply shell/platform/glfw/flutter_glfw.cc]: `u8string` isn't actually Windows-specific, so I'll leave it in :)

File: shell/platform/glfw/client_wrapper/include/flutter/flutter_window_controller.h [hunk 4]
```diff
@@ -57,17 +57,20 @@ class FlutterWindowController : public PluginRegistry {
   // Creates and displays a window for displaying Flutter content.
   //
   // The |assets_path| is the path to the flutter_assets folder for the Flutter
-  // application to be run. |icu_data_path| is the path to the icudtl.dat file
-  // for the version of Flutter you are using.
```
Comment (stuartmorgan) [inline shell/platform/glfw/client_wrapper/include/flutter/flutter_window_controller.h:61]: This should be moved up to the constructor, rather than removed.

File: shell/platform/glfw/client_wrapper/include/flutter/flutter_window_controller.h [hunk 5]
```diff
@@ -57,17 +57,20 @@ class FlutterWindowController : public PluginRegistry {
...
+  // application to be run.
   //
   // The |arguments| are passed to the Flutter engine. See:
   // https://github.com/flutter/engine/blob/master/shell/common/switches.h for
   // for details. Not all arguments will apply to desktop.
   //
+  // The |lib_path| is the path to the libapp.so file for the Flutter
+  // application to be run. This parameter is required for AOT mode support.
```
Comment (stuartmorgan) [inline shell/platform/glfw/client_wrapper/include/flutter/flutter_window_controller.h:67]: You should document that it's okay to provide a path that doesn't exist in non-AOT mode, so that clients don't need to do mode-based switching.

File: shell/platform/glfw/client_wrapper/include/flutter/flutter_window_controller.h [hunk 6]
```diff
@@ -57,17 +57,20 @@ class FlutterWindowController : public PluginRegistry {
...
+  //
   // Only one Flutter window can exist at a time; see constructor comment.
   bool CreateWindow(const WindowProperties& window_properties,
                     const std::string& assets_path,
-                    const std::vector<std::string>& arguments);
+                    const std::vector<std::string>& arguments,
+                    const std::string& lib_path = "");
```
Comment (stuartmorgan) [inline shell/platform/glfw/client_wrapper/include/flutter/flutter_window_controller.h:73]: The ordering here is unfortunate, especially since `arguments` should be removed at some point, but I guess that's something we can address later with a `DartProject` refactoring.

File: shell/platform/glfw/flutter_glfw.cc [hunk 7]
```diff
@@ -584,7 +584,10 @@ static bool RunFlutterEngine(
       std::filesystem::u8path(engine_properties.assets_path);
   std::filesystem::path icu_path =
       std::filesystem::u8path(engine_properties.icu_data_path);
-  if (assets_path.is_relative() || icu_path.is_relative()) {
+  std::filesystem::path lib_path =
+      std::filesystem::u8path(engine_properties.lib_path);
+  if (assets_path.is_relative() || icu_path.is_relative() ||
+      lib_path.is_relative()) {
```
Comment (stuartmorgan) [inline shell/platform/glfw/flutter_glfw.cc:619]: An empty path is relative, so this isn't going to work correctly. Similarly, the resolving below against `executable_location` will not give you the right outcome if the path is empty. See the Windows PR's version of the logic for this part of the code.

File: shell/platform/glfw/public/flutter_glfw.h [hunk 8]
```diff
@@ -45,6 +45,10 @@ typedef struct {
   // This can either be an absolute path, or on Windows or Linux, a path
   // relative to the directory containing the executable.
   const char* icu_data_path;
+  // The path to the libapp.so file for the application to be run.
+  // This can either be an absolute path, or on Windows or Linux, a path
+  // relative to the directory containing the executable.
```
Comment (stuartmorgan) [inline shell/platform/glfw/public/flutter_glfw.h:50]: This argument is currently only applicable to Windows and Linux, so I would rephrase this accordingly.

File: shell/platform/glfw/client_wrapper/flutter_engine.cc [hunk 9]
```diff
@@ -15,7 +15,8 @@ FlutterEngine::~FlutterEngine() {}
 
 bool FlutterEngine::Start(const std::string& icu_data_path,
                           const std::string& assets_path,
-                          const std::vector<std::string>& arguments) {
+                          const std::vector<std::string>& arguments,
+                          const std::string& lib_path) {
```
Comment (stuartmorgan) [inline shell/platform/glfw/client_wrapper/flutter_engine.cc:19]: For all the APIs, how about `aot_library_path` so that it's clearer what it is.
